python/across_api/auth/api.py

```diff
@@ -0,0 +1,90 @@
+# Copyright © 2023 United States Government as represented by the
+# Administrator of the National Aeronautics and Space Administration.
+# All Rights Reserved.
+
+
+import os
+from typing import Optional
+
+import jwt
+import requests
+from fastapi import Depends, HTTPException, Request, status
+from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
+
+from ..base.api import app
+from .schema import VerifyAuth
+
+
+class JWTBearer(HTTPBearer):
+    token_endpoint: str
+    jwks_uri: str
+
+    def __init__(self, **kwargs):
+        # Configure URL from IdP
+        user_pool_id = os.environ.get("COGNITO_USER_POOL_ID")
+        if user_pool_id is not None:
+            provider_url = f"https://cognito-idp.{user_pool_id.split('_')[0]}.amazonaws.com/{user_pool_id}/"
+        elif os.environ.get("ARC_ENV") == "testing":
+            provider_url = f"http://localhost:{os.environ.get('ARC_OIDC_IDP_PORT')}/"
+        else:
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail="Environment variable COGNITO_USER_POOL_ID must be defined in production.",
+            )
+
+        # Fetch the well-known config from the IdP
+        resp = requests.get(provider_url + ".well-known/openid-configuration")
+        try:
+            resp.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            HTTPException(status_code=500, detail=str(e))
+
+        # Find the token endpoint and jwks_uri from the well-known config
+        well_known = resp.json()
+        self.token_endpoint = well_known["token_endpoint"]
+        self.jwks_uri = well_known["jwks_uri"]
+        self.token_alg = well_known["id_token_signing_alg_values_supported"]
+
+        super().__init__(**kwargs)
+
+    async def __call__(self, request: Request) -> None:
+        credentials: Optional[HTTPAuthorizationCredentials] = await super().__call__(
+            request
+        )
+        """Validate credentials if passed"""
+        if credentials:
```

Comment:
What if `credentials` is falsey? Shouldn't we raise an exception?

Reply:
I'm pretty sure this cannot happen. If you don't provide authentication token in the header of the API code, then it will respond with `{"detail":"Not authenticated"}`. So I think this gets trapped before we even get to this code. 

That leads to the question, why have `if credentials` at all, if the code will never execute if credentials is `False` or `None`? So that's a `mypy` thing, the return type of HTTPBearer.__call__ can be None, so you have to trap that, or else mypy complains that credentials could possibly be a `None`, even if there's no way it can be. `mypy` often forces you to put extra checks in like this, even if they can never be true.

Reply:
It's of course trivial to add `else: Exception ...` at the bottom here, and I did have that originally in there, but I took it out after I realized I could not figure out a way to trigger it in my testing.

Reply:
OK! Added in 5908c29.